### ISSUE TITLE
Disable `wait` for unittest

### ISSUE THREAD

**Issue by DanEEStar** [jd/tenacity#106]
I am using the `@retry` decorator on a function which makes an HTTP-request.

I am using this exact decorator call as an example:

```
@retry(stop=stop_after_attempt(7), wait=wait_random_exponential(multiplier=1, max=60))
def func():
   ...
   requests.post(...)
```

I have a unit tests which tests, that `func` does indeed get called multiple times when the post-request fails.

But it is a bit annoying that the test takes a long time.

Is it possible to disable the `wait` time somehow only in the unit test to make the test faster?
Do I have to mock a specific function?

I also posted this on stackoverflow, in case you want some points :)
https://stackoverflow.com/questions/47906671/python-retry-with-tenacity-disable-wait-for-unittest

**Comment by immerrr** [comment]
@DanEEStar the way I handle time at least in synchronous tests is that I use a datetime-mock library, my preferred one is [freezegun](https://github.com/spulec/freezegun) but feel free to use any other. The trick is that there's usually a function to advance the frozen clock manually, for `freezegun` it is
```
frozen_datetime.tick(delta=datetime.timedelta(seconds=10))
```
With that in mind I mock `time.sleep` function not to wait for wallclock time to elapse but rather to advance the frozen clock. This kills two birds with one stone: sleep returns right away and the tests don't get stuck for a long time, but you also get the time invariant back, that is after you invoke `time.sleep`  `datetime.now()` returns a timestamp in the future. It could be slightly more involved for asynchronous tasks where the time can affect some sort of an event loop, but it's not impossible.

**Comment by davidscolgan** [comment]
If anyone finds this now, I dug into @steveb's code and found that in a later pull request it looks like you found an even simpler way of doing this:

https://github.com/openstack/tripleo-common/commit/eba4912fc4cd37442ce4603544bb070612a1df74#diff-f02c888d3d91763350bb6b11c12bf0ffR167

In short, if you decorate a function `func` with `@retry`, in the test you can do:

`func.retry.sleep = mock.Mock()`

And that makes the retry calls happen immediately.

I whipped up a gist that demonstrates Steve's method:

https://gist.github.com/davidscolgan/39433d2de29ea1282bbecaf5afd73900

**Comment by Gatsby-Lee** [comment]
@davidscolgan Thank you

**Comment by dtantsur** [comment]
> func.retry.sleep = mock.Mock()

Unfortunately, you don't always have an access to the function in a unit test. I'd rather go with the proposal in #228.

**Comment by richtier** [comment]
I prefer this apporach too `@mock.patch("tenacity.nap.time.sleep", MagicMock())` as that will auto-tear down after the test is finished, but monkey patching `func.retry.sleep = mock.Mock()` does not not tear down the monkey-patch after the test is complete.

**Comment by skinitimski** [comment]
> You can change the wait function temporarily in your test:
> 
> ```python
> func.retry.wait = wait_none
> ```

This approach no longer works after https://github.com/jd/tenacity/pull/479. I am instead trying to use the solution enabled by https://github.com/jd/tenacity/pull/236 but not having much luck yet.

**Comment by robfraz** [comment]
Yes, I'm also hitting 

> > You can change the wait function temporarily in your test:
> > ```python
> > func.retry.wait = wait_none
> > ```
> 
> This approach no longer works after #479. I am instead trying to use the solution enabled by #236 but not having much luck yet.

Yes, I'm also hitting this problem.

**Comment by thomas-dufour** [comment]
I was hitting this problem too but thanks to (https://stackoverflow.com/a/73890688) I managed to avoid this problem by patching tenacity nap directly. 
`mock.patch("tenacity.nap.time")`

**Comment by ikbenale** [comment]
In my case since my function was async I had to `@mock.patch("asyncio.sleep")`

**Comment by jmehnle** [comment]
My application talks to a variety of other services through (async) client libraries, and they each use a dedicated `Retrying` (or `AsyncRetrying`) object, so I need a way to _globally_ disable sleep during unit tests rather than patching each individual `Retrying` (or `AsyncRetrying`) object. Alas, currently there is no such way. In particular, monkey-patching `asyncio.sleep` (and, in my app's case, `trio.sleep`) is not acceptable, because this function serves other important purposes.

What I ended up doing was to:
1. define a `_sleep` function that's passed to the `sleep` argument of all my (`Async`)`Retrying` instances.
2. have this `_sleep` function check a global `_sleep_enabled` variable (statically set to `True`), and call the actual sleep (in my case `tenacity.asyncio._portable_async_sleep`) only if that's `True`.
3. define a `no_retry_sleep` context manager that temporarily sets the `_sleep_enabled` global to `False`, `yield`s, and finally restores it to `True`.

See https://gist.github.com/jmehnle/2cf4deba48c32c01fdf6c0ee5248a1d1.

Now I can use that `no_retry_sleep` context manager in my `pytest` tests or fixtures to disable sleep.

I think adding a native facility to `tenacity` for disabling sleep could be useful, but that's also possibly an incomplete solution as when you use retry with settings like `stop=stop_after_delay(600)`, then unit tests that actually exercise retry may still take 10 minutes to complete. So you'd also have to override the `stop` setting to something like `stop_after_attempt(2)`. So making `tenacity` behave well during unit tests is nontrivial and strongly depends on your actual retry configuration.

**Comment by mastizada** [comment]
I used this in my conftest to disable waiting for all my retry cases:
```python
tenacity.wait.wait_exponential_jitter.__call__ = lambda __, ___: 0
```

**Comment by JackySky** [comment]
Unfortunate that there still isn't an official solution to this. As of now the `Retrying` context is no longer exposed in the original function so the original workaround no longer work. Few approaches we could do here:  
1. You can patch `asyncio.sleep` directly but this will change _all_ `asyncio.sleep` in _anywhere_ and is not favorable when my test involve some basic timing test.
2. The approach which I ended up with is to patch the `__init__` default arguments to force for unit testing:

`patch_tenacity.py`
```
import asyncio
import tenacity.asyncio

sleep_mock = lambda: seconds: asyncio.sleep(0.1)
async def custom_sleep(seconds: float):
    return await sleep_mock(seconds)

def patch_tenacity():
    args = list(tenacity.asyncio.AsyncRetrying.__init__.__defaults__)
    args[0] = custom_sleep
    tenacity.asyncio.AsyncRetrying.__init__.__defaults__ = tuple(args)
```
And in the test file:
```
from your.module.tenacity_patch import patch_tenacity
patch_tenacity()

def test_something():
     #all wait set to 0.1 second

@patch('your.module.tenacity_patch.sleep_mock')
def test_something_else(sleep_mock):
    sleep_mock.side_effect = lambda seconds: asyncio.sleep(1) # set wait to 1 second for this test
```
I would imagine this would work for sync version and customizing retry as well, so that's good enough workaround for me - but I agree that it would be nice to have an official api to customize retry behaviours just to avoid these headaches.

**Comment by telvis07** [comment]
For anyone testing async functions with tenacity retries in pytest, here's what worked for me:

  **Problem:** Tests were failing with `tenacity.RetryError` instead of the original exception.

  **Solution:** Create a fixture that modifies the retry behavior:

I have a `Provider` class that does the lookup.

```python
class MyProvider(Protocol):
    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=1, max=10),
    )
    async def my_method(
        self, prompt: str
    ) -> ResponseSchema:
        ...
```

Then in the tests.
  ```python
  import pytest
  from tenacity import stop_after_attempt

  @pytest.fixture
  def provider_no_retry(mock_env):
      """Create provider with retry disabled for testing."""
      provider = MyProvider()
      # Stop after first attempt (no retries)
      provider.my_method.retry.stop = stop_after_attempt(1)
      # Re-raise original exception instead of wrapping in RetryError
      provider.my_method.retry.reraise = True
      return provider

  @pytest.mark.anyio
  async def test_handles_http_error(provider_no_retry):
      """Test that HTTP errors are raised correctly."""
      # Mock the HTTP call to raise an error
      with patch(...) as mock_client:
          mock_client.side_effect = httpx.HTTPStatusError(...)

          # Now the original exception is raised, not RetryError
          with pytest.raises(httpx.HTTPStatusError):
              await provider_no_retry.my_method("test")
```

The key is setting both `retry.stop = stop_after_attempt(1)` (to prevent retries) and `retry.reraise = True` (to avoid wrapping exceptions in RetryError).

**Comment by jd** [comment]
Well answered in the thread. Key approaches:
- `func.retry.wait = wait_none()`
- `func.retry.sleep = mock.Mock()`
- `mock.patch.object(func.retry, 'wait', wait_none())`
